include/swift/AST/Requirement.h

```diff
@@ -47,7 +47,10 @@ enum class RequirementKind : unsigned {
 /// types thereof) of a
 class Requirement {
   llvm::PointerIntPair<Type, 3, RequirementKind> FirstTypeAndKind;
-  Type SecondType;
+  union {
```

Comment:
In principle, it could. At least right now, because we have only two possibilities. Should we add more in the future, we may run into problems. Besides, all the getters are behind assertions currently, which check the kind of the requirement before accessing the elements of this union.